### ISSUE TITLE
Reproducible but incorrect Pylint: E1133:not-an-iterable in library

### ISSUE THREAD

**Issue by davetapley** [pylint-dev/pylint#9253]
### Bug description

```python
# iter_bug.py

from textual.app import App
from textual.widgets import Input


class MyApp(App[None]):
    def whatever(self):
        print([i.id for i in self.query(Input)])
```

### Command used

```shell
pylint iter_bug.py
```

### Pylint output

```python
************* Module iter_bug
iter_bug.py:7:29: E1133: Non-iterable value self.query(Input) is used in an iterating context (not-an-iterable)
```

### Expected behavior

No errors.

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.6 (main, Nov  2 2023, 02:14:33) [GCC 9.4.0]
```

### OS / Environment

Ubuntu 20.04.6 LTS

### Additional dependencies

```python
textual == 0.40.0
```

**Comment by davetapley** [comment]
Creating as requested [here](https://discuss.python.org/t/reproducible-but-incorrect-pylint-e1133-not-an-iterable-in-library/38633/3?u=davetapley). Initially reported here:
- https://github.com/Textualize/textual/discussions/3671

---

I did some 👀  myself:


I use VSCode debug with `launch.json`:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "pylint",
            "type": "python",
            "request": "launch",
            "module": "pylint",
            "justMyCode": false,
            "args": ["iter_bug.py"]
        },
```

Then break point here: https://github.com/pylint-dev/pylint/blob/a841a7d9eab654e423a115b4feaa857d5f807c88/pylint/checkers/typecheck.py#L2291-L2294

Add disabled break point here:
https://github.com/pylint-dev/astroid/blob/2d309cbb823e27d73bf695447a50f7875a852612/astroid/decorators.py#L94

Hit first break, then enable second break and continue.

Will hit `raise InferenceError(**error.args[0])` with something like:

```
error.args[0]
{'node': <Comprehension l.73 ...7bc9233d0>, 'unknown': <AssignName.i l.73 a...7bc923450>, 'assign_path': None, 'context': <astroid.context.Inf...7ade519c0>}
```

which causes `yield Uninferable`, then `inferred = Const.NoneType`,
and [here](https://github.com/pylint-dev/pylint/blob/a841a7d9eab654e423a115b4feaa857d5f807c88/pylint/checkers/typecheck.py#L2253-L2257) gives `not is_iterable(inferred,..` 😞 


But that's so deep in to `astroid` internals I have no idea why `InferenceError` is raised 🤔

**Comment by jacobtylerwalls** [comment]
Thanks for the investigation. From the source for `query`, which you linked, I noticed the use of `@overload`. We lack support for that (see pylint-dev/astroid#1015), so that's my best guess for why `Uninferable` turns up.

More specifically, though, the `not-an-iterable` check should quiet itself when Uninferable turns up. That's a trivial patch. Coming right up!

**Comment by jacobtylerwalls** [comment]
> More specifically, though, the not-an-iterable check should quiet itself when Uninferable turns up. That's a trivial patch. Coming right up!

Ah, it does that already. I didn't notice this part of your report:

> then inferred = Const.NoneType,

So really this is just a duplicate of pylint-dev/astroid#1015, unfortunately. Sorry there is no workaround at the moment!
